src/quo/components/tabs/tabs/view.cljs

```diff
@@ -72,31 +71,33 @@
            scroll-on-press?
            size
            style]}]
-  [rn/view
-   {:style (style/tab {:size             size
-                       :default-tab-size default-tab-size
-                       :number-of-items  number-of-items
-                       :index            index
-                       :style            style})}
-   [tab/view
-    {:id                  id
-     :notification-dot?   notification-dot?
-     :customization-color customization-color
-     :accessibility-label accessibility-label
-     :size                size
-     :blur?               blur?
-     :active              (= id @active-tab-id)
-     :on-press            (fn [id]
-                            (reset! active-tab-id id)
-                            (when (and scroll-on-press? @flat-list-ref)
-                              (.scrollToIndex ^js @flat-list-ref
-                                              #js
-                                               {:animated     true
-                                                :index        index
-                                                :viewPosition 0.5}))
-                            (when on-change
-                              (on-change id)))}
-    label]])
+  (let [on-press (rn/use-callback (fn [tab-id]
```

Comment:
@ajayesivan I'm seeing we're using `use-callback` incorrectly. One of the many perils of dropping down to the React layer. And I believe we'll continue to see more and more of this problem because that's a common problem in React apps too. We're no different.

React's `useCallback` will cache the function for subsequent renders relying exclusively on the deps passed to the hook. When we don't pass all the necessary dependencies, the cached function, if used again, will be incorrect.

In this code, `index` and `scroll-on-press?` are not being passed as dependencies. This code's PR only works if `data` (the sequence of tabs) doesn't change its size ever, which is a fragile assumption to be made.

Here's a proof demonstrating a bug:

1. In the Activity Center, there are 9 tabs. 0 to 8 `index`.
2. Make the tabs in `status-im.contexts.shell.activity-center.tabs.view` dynamic with `use-state`.
3. Set a new tab state where you have only 2 tabs (e.g. `Mentions` and `Membership`)
4. The UI re-rendered and now shows only two tabs.
5. Press on any of the tabs and watch the app explode with `scrollToIndex out of range`.

One might say, "but wait, you're forcing a bug, the current prod code uses static tabs". For that I say no, React components should be made dynamic and render correctly whenever state changes. Here it's not even the component's state directly, but a prop change from a parent view, so it's worse. @ajayesivan, this is by no means a feedback for your PR alone, but a general problem we'll continue to see and we need to be careful.

@flexsurfer this is a good example of a problem we didn't have with Reagent (missing deps for cached functions).

-------

Long story short, `use-callback` requires careful thought from developers. It's not always needed, just as we coded for years with Reagent without caching all functions. Particularly in components that render only once or re-render rarely I would say it's actually detrimental to the quality of the app.

cc @ulisesmac this is one good example about what we were talking today about hooks' deps.

Reply:
Devs must understand how react works, and we should have full control of it, just rerendering functions and components each time is not a good option, so reagent simply hid the problems by paying for it in performance

Reply:
> Devs must understand how react works, and we should have full control of it, just rerendering functions and components each time is not a good option, so reagent simply hid the problems by paying for it in performance

Right @flexsurfer, we should all know what we're using 👍🏼 

My point is different, it's that `use-callback` shouldn't be used all the time because caching functions in React is only ever (maybe) going to make a difference in parts of the tree that re-render more frequently. And the downside of using it is that it's very easy to use the incorrect deps and cause hard to debug issues. As I've shown in my previous comment.

In React apps we have linters to help detect incorrect deps. UIx for instance does this check at _compile-time_ (currently experimental). We have little to no protection in status-mobile.

Overall, I think we shouldn't be using `use-callback` everywhere there's an anonymous function. Caching functions is not needed all the time, far from that, because the cache will (maybe) be used only on the next render (if it happens). Our guidelines should be flexible regarding `use-callback`.

In the guidelines:

```
rn/use-callback hook should be used for anon callback functions
```

In essence, this hard rule shouldn't always be followed in my experience with React apps.

Reply:
I agree here that the guideline is too strict. Whether we use reagent or no, the problem is the same, we want to prevent recreating function across re-renders in performance critical paths, or have the function cause unnecessary re-renders because of the way is defined. 

But using `rn/use-callback` for everything sounds like a sledgehammer and I think gives a false sense of security, like in the case above, as it might be unnecessary, and I agree that it can be prone to errors (long interesting discussion to be had, I think, it also forces developers to think about dependencies of the function, which I guess is good as many of the mistakes pointed out were made because we didn't think about dependencies, but not worth the candle imo, but that's a separate discussio) . 

Unfortunately it's one of those cases were developer needs to think about the lifecycle and dependencies of their components and write the code accordingly, if they are not doing that, that's were code review and more introspection on why something re-renders and how many times will help. So I would agree that the guideline need to be changed into something that will give developers more agency on when it's good to use, and when it's not necessary.